rafx-framework/src/render_features/jobs/prepare/prepare_job.rs

```diff
@@ -0,0 +1,257 @@
+                frame_packet,
+                submit_packet,
+                view_packet,
+                view_submit_packet,
+                id,
+            );
+            self.inner.prepare_render_object_instance_per_view(&context);
+        }
+    }
+
+    fn end_per_view_prepare(
+        &self,
+        view_packet: &dyn RenderFeatureViewPacket,
+        view_submit_packet: &dyn RenderFeatureViewSubmitPacket,
+    ) {
+        profiling::scope!(PrepareScopeNames::scope(
+            &self.scope_names.end_per_view_prepare
+        ));
+
+        let view_packet: &ViewPacket<PrepareJobEntryPointsT::FramePacketDataT> =
+            view_packet.as_concrete();
+
+        let view_submit_packet: &ViewSubmitPacket<PrepareJobEntryPointsT::SubmitPacketDataT> =
+            view_submit_packet.as_concrete();
+
+        let context = PreparePerViewContext::new(
+            self.frame_packet.as_ref().unwrap(),
+            self.submit_packet.as_ref().unwrap(),
+            view_packet,
+            view_submit_packet,
+        );
+
+        if let Some(render_objects) = self.inner.render_objects() {
+            self.inner.end_per_view_prepare(&render_objects, &context);
+        } else {
+            self.inner
+                .end_per_view_prepare(&self.empty_render_objects, &context);
+        }
+    }
+
+    fn end_per_frame_prepare(&self) {
+        profiling::scope!(PrepareScopeNames::scope(
+            &self.scope_names.end_per_frame_prepare
+        ));
+
+        let context = PreparePerFrameContext::new(
+            self.frame_packet.as_ref().unwrap(),
+            self.submit_packet.as_ref().unwrap(),
+        );
+        if let Some(render_objects) = self.inner.render_objects() {
+            self.inner.end_per_frame_prepare(&render_objects, &context);
+        } else {
+            self.inner
+                .end_per_frame_prepare(&self.empty_render_objects, &context);
+        }
+    }
+
+    fn num_views(&self) -> usize {
+        self.view_packets().len()
+    }
+
+    fn num_render_object_instances(&self) -> usize {
+        self.render_object_instances().len()
+    }
+
+    fn take_frame_packet(&mut self) -> Box<dyn RenderFeatureFramePacket> {
+        std::mem::take(&mut self.frame_packet).unwrap()
+    }
+
+    fn take_submit_packet(&mut self) -> Box<dyn RenderFeatureSubmitPacket> {
+        std::mem::take(&mut self.submit_packet).unwrap()
+    }
+
+    fn feature_debug_name(&self) -> &'static str {
+        self.inner.feature_debug_name()
+    }
+
+    fn feature_index(&self) -> RenderFeatureIndex {
+        self.inner.feature_index()
+    }
+}
+
+struct PrepareScopeNames {
+    begin_per_frame_prepare: String,
+    prepare_render_object_instance: String,
+    prepare_render_object_instance_per_view: String,
+    end_per_view_prepare: String,
+    end_per_frame_prepare: String,
+}
+
+impl PrepareScopeNames {
+    pub fn new(feature_name: &str) -> Self {
+        Self {
```

Comment:
maybe we can lazy_static this

Reply:
or create these strings when we register the feature?

Reply:
Creating them when the feature is registered is an interesting idea. I can look at that.

Reply:
Dupe of https://github.com/aclysma/rafx/pull/135#discussion_r631574648, will resolve there.